### ISSUE TITLE
[Potential NAN bug] Loss may become NAN during training

### ISSUE THREAD

**Issue by Justobe** [kaneshirok/ml-plactice#1]
Hello~

Thank you very much for sharing the code!

I try to run script `src/ch5/bmi.py` . After some iterations, it was found that the training loss became NAN. After carefully checking the code, I found that the following code may trigger NAN in loss:

In `ml-plactice/src/ch5/bmi.py: line32`:

```python
cross_entropy = -tf.reduce_sum(y_ * tf.log(y))
```

If  y contains 0 (output of softmax ), the result of `tf.log(y)` is `inf` because `log(0)` is illegal . And this may cause the result of loss to become NAN.

It could be fixed by making the following changes:

```python
cross_entropy = -tf.reduce_sum(y_ * tf.log(y+1e-8))
```

or

```python
cross_entropy = -tf.reduce_sum(y_* tf.log(tf.clip_by_value(y,1e-8,1.0)))
```

The same problem was also found in `src/ch5/tb-bmi.py`, `src/ch5/tb-bmi2.py`, and `src/ch5/mnist-deep.py`

Hope to hear from you ~

Thanks in advance!  : )
